### DISCUSSION TITLE
`ref` is called before element is ready in DOM, or is it by design ?

### DISCUSSION THREAD

**Discussion by yw662** [solidjs/solid#1292] [Q&A]
### Describe the bug

like,
```tsx
<div ref={element => console.log(element.parentElement)}></div>
```
It will always be `null`, as it isn't in the DOM yet.

However if we want to do sth at the moment the given element is in DOM,  exp `getBoundingClientRect`, it can be annoying.
In some cases this can be done by using `onMount` instead, as in #1217, however `ref` can be much easier sometime, or you may have to use both.

It won't be surprising if this behavior is by design, but is there a "recommended practice" for this usage ?

### Your Example Website or App

not yet

### Steps to Reproduce the Bug or Issue

just the code above.

### Expected behavior

I hope there is a way to make it non null.

### Platform

- OS: NA
- Browser: any
- Version: 1.5.9

**Comment by ryansolid** [comment]
Yeah refs are assigned element creation time. No need to do another pass. We usually point people to use them with `onMount` or `createEffect`. Those run after everything has been attached.

**Comment by yw662** [comment]
So the recommended practice is to store the ref somewhere (into a signal or store maybe), and do the logic in a `onMount` or `createEffect`.
Is it possible to omit the `onMount` part somehow, so the code can be easier to read ?

**Comment by ryansolid** [comment]
No. I don't think so. It compiles to just calling the provided function. So this is asking if the framework can insert `onMount` because Solid otherwise doesn't touch these nodes again. The current behavior would break in this case as `ref`'s are present to be part of rendering.

**Comment by zaydek** [comment]
I wonder if you could do something like this:

```tsx
<div ref={el => {
  onMount(() => {
    setClientBoundingBox(el.getBoundingClientRect())
  })
}}>
```

**Reply by yw662** [reply]
It looks good but I wonder whether it works in all scenarios.
